feat: new blog post about starting a workspace from URL (#13)
eclipse-che/blog: _posts/2021-12-20-authoring-a-url-to-start-a-workspace.adoc

---
title: Authoring a URL to start a workspace
layout: post
author: Florent Benoit
description: >-
  How to start a workspace by customizing the URL.
categories: []
keywords: ['GitHub', 'GitLab', 'Bitbucket', 'start', 'workspace']
slug: /@florent.benoit/start-a-workspace-by-entering-a-url
---

> // https://excalidraw.com/#json=Z4gT6H3GoZwJkfvVCajad,Qtltwy9uKpQOtosF8mh00g
> image::../../../assets/img/start-a-workspace/start-workspace-link.png[Start a workspace link,link="../../assets/img/start-a-workspace/start-workspace-link.png"]
> 
> == Authoring a URL to start a workspace
> 
> As a user, you can create a workspace containing a clone of a Git repository by entering:
> 
> * a custom URL in the address bar of your browser.
> * the URL of the Git repository on the {prod} dashboard page.
> 

Reviewer: ```suggestion

== Authoring a URL to start a workspace

As a user, you can create a workspace using a direct URL rather than using the user dashboard. 
This blog post describes how to author a URL to start a workspace.

To use a URL to start a workspace you can choose to:

* Paste the URL in the address bar of your browser.
* Distribute the URL as a link (in an onboarding procedure, in a pull request comment).

// https://excalidraw.com/#json=Z4gT6H3GoZwJkfvVCajad,Qtltwy9uKpQOtosF8mh00g
image::../../../assets/img/start-a-workspace/start-workspace-link.png[Authoring a URL to start a workspace,link="../../assets/img/start-a-workspace/start-workspace-link.png"]
```

Reviewer: * Decouple authoring the URL / using the URL.
* Add the image after the lead in sentence (it has more context).

.Mandatory repository parameter:
[options="header"]
|=======================
|parameter|Description|example
| `#__<repository_URL>__` | Repository URL | `#https://github.com/eclipse-che/che-server/tree/7.40.x`
|=======================

.Optional parameters: (`?` followed by `keyword=value`)
[options="header"]
|=======================
|Keyword |Description |Example
|`che-editor=__<ide>__`| IDE | `?che-editor=eclipse/che-theia/latest`
|`devfilePath=__<devfile>__`| Devfile path | `?devfilePath=.che/my-devfile.yaml`
|`new`| Always create a new workspace|`?new`
|=======================
Note: `?new` keyword has no value

In the last part of the URL, parameter list preceded by ? and using the & separator.

__<{prod-url}>__#__<repository_url>__?__<parameter_1>__&__<parameter_2>__

> === Defining the Git repository to clone

Reviewer: ```suggestion
=== Defining the Git repository URL
```

more consistent with the rest of the post

> Short syntax is available when using {devworkspace} Operator.

Reviewer: ```suggestion
This syntax is available when using {devworkspace} Operator.
```

> To start a workspace with a clone of a GitHub, GitLab, or Bitbucket repository, go to your browser's address bar and enter the {prod-short} URL, prepended with a `#` character and appended with the repository link. Using the example of Eclipse Che hosted by Red Hat: link:https://workspaces.openshift.com#https://github.com/eclipse-che/che-server[]. Then {prod} starts a workspace and clones the repository, `\https://github.com/eclipse-che/che-server` in this example.

Reviewer: ```suggestion
The Git repository URL is mandatory.
{prod-short} supports following Git providers: GitHub, GitLab, or on-premise Bitbucket.

Add a `#` character and append the repository URL: `#__<repository_URL>__`.

.Open Eclipse Che Server repository in Eclipse Che hosted by Red Hat
====
link:https://workspaces.openshift.com#https://github.com/eclipse-che/che-server[].
====

Then {prod} starts a workspace and clones the repository, `\https://github.com/eclipse-che/che-server` in this example.
```

---
title: Authoring a URL to start a workspace
layout: post
author: Florent Benoit
description: >-
  How to start a workspace by customizing the URL.
categories: []
keywords: ['GitHub', 'GitLab', 'Bitbucket', 'start', 'workspace']
slug: /@florent.benoit/authoring-a-url-to-start-a-workspace
---

== Authoring a URL to start a workspace

As a user, you can create a workspace using a direct URL rather than navigating the user dashboard. 
This blog post describes how to author a URL to start a workspace.

To use a URL to start a workspace you can choose to:

* Paste the URL in the address bar of your browser.
* Distribute the URL as a link (in an onboarding procedure, in a pull request comment).

> 

Reviewer: ```suggestion

When you click on a URL to start a workspace, {prod}:

. Creates or reuses a workspace.
. Clones the repository.
. Searches for a devfile in the given repository. A parameter can override the default location: `.devfile.yaml` or `devfile.yaml`. The devfile describes the containers and commands to use for that project. See the link:https://devfile.io/[Devfile Documentation]. Omitting this file will result in a default container, which provides different languages (including Java, Node.js, Python, PHP, Golang, etc) and tools (including `curl`, `jq`, `yq`, etc).
. Starts the containers to use for that project.
. Starts the IDE. A parameter can override the default IDE.

```

|=======================
|parameter|Description|example
| `#__<repository_URL>__` | Repository URL | `#https://github.com/eclipse-che/che-server/tree/7.40.x`
|=======================

.Optional parameters: (`?` followed by `keyword=value`)
[options="header"]
|=======================
|Keyword |Description |Example
|`che-editor=__<ide>__`| IDE | `?che-editor=eclipse/che-theia/latest`
|`devfilePath=__<devfile>__`| Devfile path | `?devfilePath=.che/my-devfile.yaml`
|`new`| Always create a new workspace|`?new`
|=======================
Note: `?new` keyword has no value

In the last part of the URL, parameter list preceded by ? and using the & separator.

__<{prod-url}>__#__<repository_url>__?__<parameter_1>__&__<parameter_2>__

=== Defining the Git repository URL

This syntax is available when using {devworkspace} Operator.

The Git repository URL is mandatory.
{prod-short} supports following Git providers: GitHub, GitLab, or on-premise Bitbucket.

Add a `\#` character and append the repository URL: `#__<repository_URL>__`.

.Open Eclipse Che Server repository in Eclipse Che hosted by Red Hat
====
link:https://workspaces.openshift.com#https://github.com/eclipse-che/che-server[].
====

> Then {prod} starts a workspace and clones the repository, `\https://github.com/eclipse-che/che-server` in this example.
> 
> 
> {prod} also searches for a `.devfile.yaml` or `devfile.yaml` file in the given repository. The devfile describes the containers and commands to use for that project. Omitting this file will result in a default container, which provides different languages (including Java, Node.js, Python, PHP, Golang, etc) and tools (including `curl`, `jq`, `yq`, etc).
> 
> For more information about devfiles, see the link:https://devfile.io/[Devfile Documentation].

Reviewer: ```suggestion
When you click on a link containing a Git repository URL, {prod}:

. Starts a workspace.
. Clones the repository, `\https://github.com/eclipse-che/che-server` in this example.
. Searches for a `.devfile.yaml` or `devfile.yaml` file in the given repository. The devfile describes the containers and commands to use for that project. See the link:https://devfile.io/[Devfile Documentation]. Omitting this file will result in a default container, which provides different languages (including Java, Node.js, Python, PHP, Golang, etc) and tools (including `curl`, `jq`, `yq`, etc).
. Starts the containers to use for that project.

```

---
title: Authoring a URL to start a workspace
layout: post
author: Florent Benoit
description: >-
  How to start a workspace by customizing the URL.
categories: []
keywords: ['GitHub', 'GitLab', 'Bitbucket', 'start', 'workspace']
slug: /@florent.benoit/authoring-a-url-to-start-a-workspace
---

As a user, you can create a workspace using a direct URL rather than navigating the user dashboard. 
This blog post describes how to author a URL to start a workspace.

> 

Reviewer: ```suggestion
The coordinates are organized into the standard sequence shown below.
```

// https://excalidraw.com/#json=Z4gT6H3GoZwJkfvVCajad,Qtltwy9uKpQOtosF8mh00g
image::../../../assets/img/start-a-workspace/start-workspace-link.png[Authoring a URL to start a workspace,link="../../assets/img/start-a-workspace/start-workspace-link.png"]

To use a URL to start a workspace you can choose to:

* Paste the URL in the address bar of your browser.
* Distribute the URL as a link (in an onboarding procedure, in a pull request comment).

When you click on a URL to start a workspace, {prod}:

. Creates or reuses a workspace. A parameter can request to always create a new workspace.
. Clones the repository.
. Searches for a devfile in the given repository. A parameter can override the default location: `.devfile.yaml` or `devfile.yaml`. The devfile describes the containers and commands to use for that project. See the link:https://devfile.io/[Devfile Documentation]. Omitting this file will result in a default container, which provides different languages (including Java, Node.js, Python, PHP, Golang, etc) and tools (including `curl`, `jq`, `yq`, etc).
. Starts the containers to use for that project.
. Starts the IDE. A parameter can override the default IDE.

> 

Reviewer: ```suggestion

== URL parameters reference

```

====

* To clone a specified branch on GitHub to the starting workspace:
+
`#https://github.com/_<user_or_org>_/_<repository>_/tree/_<branch-name>_`
+
.Cloning the `7.39.x` branch of the link:https://github.com/eclipse-che/che-plugin-registry[Plug-in registry repository].
====
`#https://github.com/eclipse-che/che-plugin-registry/tree/7.39.x`
====

* To clone the branch of a pull request on GitHub to the starting workspace:
+
`#https://github.com/_<user_or_org>_/_<repository>_/pull/_<pull-request-id>_`
+
.Cloning the link:https://github.com/eclipse-che/che-plugin-registry/pull/1097[Plug-in registry repository pull request `#1097`].
====
`#https://github.com/eclipse-che/che-plugin-registry/pull/1097`
====

URL syntax for cloning GitLab-hosted repositories:

* To clone the default branch on GitLab to the starting workspace:
+
`#https://gitlab.com/_<user_or_org>_/_<repository>_`

* To clone a specified branch on GitLab to the starting workspace:
+
`#https://gitlab.com/_<user_or_org>_/_<repository>_/-/tree/_<branch>_`

URL syntax for cloning BitBucket-hosted repositories:

WARNING: {prod} supports only on premises BitBucket instances.

> === Defining the editor or Integrated Development Environment (IDE)

Reviewer: ```suggestion
== Defining the editor or Integrated Development Environment (IDE)
```

{prod} starts workspaces with a default IDE: {prod-short} Theia.

Overriding the default editor is possible with the `?che-editor=__<editor-id>__` parameter. The `editor-id` comes from the identifier of the IDE in the plug-in registry.

.Using the next version of Theia (build after each commit) 
====
`pass:c,a,q[{prod-url}#https://github.com/example/foo?che-editor=eclipse/che-theia/next]`
====

Some available IDE keys:

* `eclipse/che-theia/latest`
* `eclipse/che-theia/next`
* `che-incubator/intellij-community/latest`
* `che-incubator/pycharm/latest` 
* `che-incubator/che-code/insiders`

Obtain IDE keys from the plug-in registry by looking at `Che Editor` type entries.
====
`pass:c,a,q[{prod-url}/che-plugin-registry/main/index.json]`
====

The parameter `?che-editor` accepts a link to the devfile hosted on a public server (with Cross-Origin Resource Sharing (CORS) enabled).

====
`pass:c,a,q[{prod-url}#https://github.com/example/foo?che-editor=https://eclipse-che.github.io/che-plugin-registry/main/v3/plugins/che-incubator/che-code/insiders/devfile.yaml]`
====

> === Using a different devfile file from the repository

Reviewer: ```suggestion
== Using a different devfile file from the repository
```

If the devfile file name is different from the default search pattern, which is `.devfile.yaml` or `devfile.yaml`, it is possible to override that name.

Use the parameter `?devfilePath=mydevfile.yaml` at the end of `{prod-url}#__<git_repository_url>__`.

====
`pass:c,a,q[{prod-url}#https://github.com/example/foo?devfilePath=mydevfile.yaml]`
====

The short name for the `devfilePath` parameter is `df`.

====
`pass:c,a,q[{prod-url}#https://github.com/example/foo?df=mydevfile.yaml]`
====

> === Creating or reusing a workspace

Reviewer: ```suggestion
== Creating or reusing a workspace
```

If you enter a URL such as `pass:c,a,q[{prod-url}#http://github.com/foo/bar]` and you don’t have any workspace named `bar`, {prod} creates a workspace named `bar` and clones the GitHub repository.

Each time you enter a URL such as `{prod-url}#http://github.com/foo/bar` again, {prod} reopens the existing workspace `bar`.

> === Always creating a new workspace

Reviewer: ```suggestion
== Always creating a new workspace
```

---
title: Authoring a URL to start a workspace
layout: post
author: Florent Benoit
description: >-
  How to start a workspace by customizing the URL.
categories: []
keywords: ['GitHub', 'GitLab', 'Bitbucket', 'start', 'workspace']
slug: /@florent.benoit/authoring-a-url-to-start-a-workspace
---

== Authoring a URL to start a workspace

As a user, you can create a workspace using a direct URL rather than navigating the user dashboard. 
This blog post describes how to author a URL to start a workspace.

> To use a URL to start a workspace you can choose to:
> 
> * Paste the URL in the address bar of your browser.
> * Distribute the URL as a link (in an onboarding procedure, in a pull request comment).
> 
> When you click on a URL to start a workspace, {prod}:
> 
> . Creates or reuses a workspace. A parameter can request to always create a new workspace.
> . Clones the repository.
> . Searches for a devfile in the given repository. A parameter can override the default location: `.devfile.yaml` or `devfile.yaml`. The devfile describes the containers and commands to use for that project. See the link:https://devfile.io/[Devfile Documentation]. Omitting this file will result in a default container, which provides different languages (including Java, Node.js, Python, PHP, Golang, etc) and tools (including `curl`, `jq`, `yq`, etc).
> . Starts the containers to use for that project.
> . Starts the IDE. A parameter can override the default IDE.
> 
> // https://excalidraw.com/#json=Z4gT6H3GoZwJkfvVCajad,Qtltwy9uKpQOtosF8mh00g
> image::../../../assets/img/start-a-workspace/start-workspace-link.png[Authoring a URL to start a workspace,link="../../assets/img/start-a-workspace/start-workspace-link.png"]

Reviewer: ```suggestion

// https://excalidraw.com/#json=Z4gT6H3GoZwJkfvVCajad,Qtltwy9uKpQOtosF8mh00g
image::../../../assets/img/start-a-workspace/start-workspace-link.png[Authoring a URL to start a workspace,link="../../assets/img/start-a-workspace/start-workspace-link.png"]

To use a URL to start a workspace you can choose to:

* Paste the URL in the address bar of your browser.
* Distribute the URL as a link (in an onboarding procedure, in a pull request comment).

When you click on a URL to start a workspace, {prod}:

. Creates or reuses a workspace. A parameter can request to always create a new workspace.
. Clones the repository.
. Searches for a devfile in the given repository. A parameter can override the default location: `.devfile.yaml` or `devfile.yaml`. The devfile describes the containers and commands to use for that project. See the link:https://devfile.io/[Devfile Documentation]. Omitting this file will result in a default container, which provides different languages (including Java, Node.js, Python, PHP, Golang, etc) and tools (including `curl`, `jq`, `yq`, etc).
. Starts the containers to use for that project.
. Starts the IDE. A parameter can override the default IDE.

```

---
title: Authoring a URL to start a workspace
layout: post
author: Florent Benoit
description: >-
  How to start a workspace by customizing the URL.
categories: []
keywords: ['GitHub', 'GitLab', 'Bitbucket', 'start', 'workspace']
slug: /@florent.benoit/authoring-a-url-to-start-a-workspace
---

== Authoring a URL to start a workspace

As a user, you can create a workspace using a direct URL rather than navigating the user dashboard. 
This blog post describes how to author a URL to start a workspace.

To use a URL to start a workspace you can choose to:

* Paste the URL in the address bar of your browser.
* Distribute the URL as a link (in an onboarding procedure, in a pull request comment).

When you click on a URL to start a workspace, {prod}:

> . Creates or reuses a workspace.

Reviewer: ```suggestion
. Creates or reuses a workspace. A parameter can request to always create a new workspace.
```

|=======================
|parameter|Description|example
| `#__<repository_URL>__` | Repository URL | `#https://github.com/eclipse-che/che-server/tree/7.40.x`
|=======================

.Optional parameters: (`?` followed by `keyword=value`)
[options="header"]
|=======================
|Keyword |Description |Example
|`che-editor=__<ide>__`| IDE | `?che-editor=eclipse/che-theia/latest`
|`devfilePath=__<devfile>__`| Devfile path | `?devfilePath=.che/my-devfile.yaml`
|`new`| Always create a new workspace|`?new`
|=======================
Note: `?new` keyword has no value

In the last part of the URL, parameter list preceded by ? and using the & separator.

__<{prod-url}>__#__<repository_url>__?__<parameter_1>__&__<parameter_2>__

=== Defining the Git repository URL

This syntax is available when using {devworkspace} Operator.

The Git repository URL is mandatory.
{prod-short} supports following Git providers: GitHub, GitLab, or on-premise Bitbucket.

Add a `\#` character and append the repository URL: `#__<repository_URL>__`.

.Open Eclipse Che Server repository in Eclipse Che hosted by Red Hat
====
link:https://workspaces.openshift.com#https://github.com/eclipse-che/che-server[].
====

> When you click on a link containing a Git repository URL, {prod}:
> 
> . Starts a workspace.
> . Clones the repository, `\https://github.com/eclipse-che/che-server` in this example.
> . Searches for a `.devfile.yaml` or `devfile.yaml` file in the given repository. The devfile describes the containers and commands to use for that project. See the link:https://devfile.io/[Devfile Documentation]. Omitting this file will result in a default container, which provides different languages (including Java, Node.js, Python, PHP, Golang, etc) and tools (including `curl`, `jq`, `yq`, etc).
> . Starts the containers to use for that project.

Reviewer: ```suggestion
```
Moved up
